openedx/core/djangoapps/plugins/plugin_slots.py

```diff
@@ -0,0 +1,106 @@
+from importlib import import_module
+from logging import getLogger
+
+from importlib_metadata import ModuleNotFoundError
+
+from openedx.core.lib.cache_utils import process_cached
+from . import constants, registry
+from .plugin_contexts import get_plugins_view_context
+
+log = getLogger(__name__)
+
+COMMON_WHITELIST = ['request', 'current_url']
+
+
+def get_content_for_slot(project_type, slot_name, raw_context):
+    """
+    Returns a list of additional content for a view slot. Will check if any plugin apps
+    have that view in their view_context_config, and if so will call their
+    selected function to get their context dicts.
+    Params:
+        project_type: a string that determines which project (lms or studio) the view is being called in. See the
+            ProjectType enum in plugins/constants.py for valid options
+        slot_name: a string that determines which slot the plugin will render content for. These are unique for
+            each project type.
+    """
+    aggregate_slot_content = u""
+    slot_functions = _get_cached_slot_functions_for_view(project_type, slot_name)
+
+    # Each view can pass the whitelisted part of the context, in the context itself.
+    context_whitelist = raw_context.get('context_whitelist', [])
+    if context_whitelist == '*':
+        whitelisted_context = raw_context
+    else:
+        # The request object is whitelisted by default, and always available.
+        context_whitelist.extend(COMMON_WHITELIST)
+
+        whitelisted_context = {
+            key: raw_context.get(key)
+            for key in context_whitelist
+        }
+
+    for (slot_function, plugin_name) in slot_functions:
+        try:
+            # Allow plugins to extend the context for other plugins
+            context = get_plugins_view_context(
+                constants.ProjectType.LMS,
+                plugin_name,
+                raw_context,
+            )
+            context.update(whitelisted_context)
+            plugin_slot_content = slot_function(context)
+            aggregate_slot_content += plugin_slot_content
```

Comment:
So is this effectively concatenating HTML strings together based on the declaration order of plugins for a given slot?  Is there facility in the `PluginSlots.CONFIG` to provide an array of plugins for a given slot?  I'm possibly missing something here.

Reply:
@davidjoy I don't think there is any expected order. We're iterating `registry.get_app_configs` which in turn is iterating over the values of the plugin dictionary. Since it's a dictionary I don't think there is any defined order in Python < 3.6. 

Would providing a list of plugins for a given slot help much? Whoever codes a plugin can just handle that internally.

Reply:
If multiple plugins want to insert content in the same slot, are we saying the order in which that content gets inserted is non-deterministic?  I think that's my concern.  But I suppose plugins shouldn't be relying on each other anyway.

Reply:
There are a number of different solutions here, tell me what makes sense:

1. Allow plugins to specify a priority/order. This can be just another entry in their config. 
2. Allow users to specify a priority/order in the config (lms.yaml). Users can specify a plugin order in the config file, this can either be a list of plugins, in which case the listed plugins will be run in the order they appear on that list, and if some don't they will not be run/ be run at the end. 
3. Allow users to specify the config in a django model (perhaps siteconfiguration?). Similar to above, but the config is stored in the database. 

In the absence of a config the order can be non-deterministic, or we can just sort the list so it's deterministic at least. The configuration specifying a priority list could be something that applies globally, or at a per-slot level, i.e. for the same three plugins, the order can be different depending on which slot they are being used for